---
title: "Tutorial 2. Pre-processing snRNA-seq"
author: "Iguaracy Souza and Umran YAMAN"
date: 'Compiled: `r Sys.Date()`'  
output: 
  html_document:
    toc: TRUE
    code_folding: show
    theme: united
    df_print: kable
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  warning = FALSE,
  error = TRUE,
  echo = TRUE,
  message = FALSE,
  fig.width = 8,
  time_it = TRUE,
  include = TRUE
)
```

### Background: 

Recent studies have identified numerous non-coding loci associated with Alzheimer’s disease (AD) risk, yet their underlying mechanisms and the transcriptional regulatory circuitry specific to AD remain poorly elucidated. In this investigation, we conducted a comprehensive analysis of the epigenomic and transcriptomic landscapes in 850,000 nuclei derived from AD and healthy prefrontal cortex tissues of 92 individuals. Our aim was to construct a detailed map of the brain regulome, encompassing epigenomic profiles, transcriptional regulators, co-accessibility modules, and peak-to-gene links in a cell-type-specific context. To achieve this, we devised novel methods for multimodal integration and the identification of regulatory modules through peak-to-gene linking. Our findings reveal an enrichment of AD risk loci in microglial enhancers and implicate specific transcription factors, such as SPI1, ELF2, and RUNX1. Additionally, we identified 9,628 cell-type-specific ATAC-QTL loci, which, when integrated with peak-to-gene links, allowed us to prioritize regulatory circuits associated with AD variants. Notably, we observed differential accessibility in regulatory modules, with glia showing changes in late-stage AD and neurons exhibiting alterations in early-stage AD. Strikingly, late-stage AD brains displayed a global dysregulation of the epigenome, indicative of epigenomic erosion and loss of cell identity.

### Dataset

Load the downsampled RNA-seq-seq dataset, which could be found in the data directory. The dataset consists 0.05% of each cell type proportion out of 414k scRNA-seq dataset which could be found [here](https://compbio.mit.edu/ad_epigenome/). The paper is [Xiong et al. (2023)](https://www.sciencedirect.com/science/article/pii/S0092867423009741?via%3Dihub) . This dataset is originally downsampled per cell type proportion (0.05%), and the original file can be found on the link [RNA.h5ad](https://personal.broadinstitute.org/bjames/AD_snATAC/RNA/)


Load the required libraries for the quality control steps of the scRNA-seq dataset

```{r}
suppressPackageStartupMessages({
  library("zellkonverter")
  library("ggplot2")
  library("reticulate")
  library("SingleCellExperiment")
  library("AnnotationDbi")
  library("org.Hs.eg.db")
  library("EnsDb.Hsapiens.v86")
  library("scater")
  library("Matrix")
  library("Seurat")
  library("DoubletFinder")
  library("dplyr")
})
```


Load the dataset using readRDS function. The rds file can be found on the dropbox link above. 

```{r}
scRNA_brain_seurat_QC <- readRDS("../data/scRNA_brain_seurat_downsampled.rds")
DefaultAssay(scRNA_brain_seurat_QC) <- "RNA"
```

### Quality Control

The quality control involves several steps to ensure the quality/reliability of the data. 

Initially, we can generate visualisations across selected metadata features whether there are outliers, and assess correlations between different features, subsetting and cleaning the data based on specific criteria, calculating cell cycle scores, identifying variable features, performing data scaling, running PCA for dimensionality reduction, and checking metadata for potential doublets.

### Visualisations of selected metadata features. 


```{r}
VlnPlot(scRNA_brain_seurat_QC, features = c("n_genes","n_counts","pct_mito","pct_ribo"), ncol = 4,pt.size = 0.1) +  
  theme(plot.title = element_text(size=10))
```



Let’s plot some of the metadata features against each other and see how they correlate. The number above each plot is a Pearson correlation coefficient.


```{r}
FeatureScatter(scRNA_brain_seurat_QC, feature1 = "n_counts", feature2 = "pct_mito")
```


```{r}
FeatureScatter(scRNA_brain_seurat_QC, feature1 = "n_counts", feature2 = "n_genes")
```


```{r}
FeatureScatter(scRNA_brain_seurat_QC, feature1 = "n_counts", feature2 = "pct_ribo")
```


```{r}
FeatureScatter(scRNA_brain_seurat_QC, feature1 = "pct_ribo", feature2 = "pct_mito")
```

```{r}
##subset scRNA_brain_seurat
subset(
  scRNA_brain_seurat_QC,
  n_genes>750 & 
    n_genes < 8000 & 
    pct_mito < 10 
  #percent.Largest.Gene < 10
) -> scRNA_brain_seurat_QC

scRNA_brain_seurat_QC
```


```{r}
cc.genes.updated.2019$s.genes <- UpdateSymbolList(symbols = cc.genes.updated.2019$s.genes)
cc.genes.updated.2019$g2m.genes <- UpdateSymbolList(symbols = cc.genes.updated.2019$g2m.genes)

s.genes <- cc.genes.updated.2019$s.genes
g2m.genes <- cc.genes.updated.2019$g2m.genes
cc.genes <- cc.genes.updated.2019
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
NormalizeData(scRNA_brain_seurat_QC, normalization.method = "LogNormalize", scale.factor = 10000) -> scRNA_brain_seurat_QC
scRNA_brain_seurat_QC <- CellCycleScoring(scRNA_brain_seurat_QC, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

VlnPlot(scRNA_brain_seurat_QC, features = c("S.Score", "G2M.Score"),
        ncol = 4, pt.size = 0.1)
```

```{r}
scRNA_brain_seurat_QC <- FindVariableFeatures(scRNA_brain_seurat_QC, selection.method = "vst", nfeatures = 2000, verbose = F)
top10 <- head(VariableFeatures(scRNA_brain_seurat_QC), 10)

plot1 <- VariableFeaturePlot(scRNA_brain_seurat_QC)
LabelPoints(plot = plot1, points = top10, repel = TRUE, xnudge = 0, ynudge = 0)
LabelPoints(plot = plot1, points = top10, repel = TRUE, xnudge = 0, ynudge = 0)
```


### Regress out the read depth, mitochondrial percentage and cell cycling genes

```{r}
scRNA_brain_seurat_QC = ScaleData(scRNA_brain_seurat_QC, vars.to.regress = c("n_genes", "pct_mito","S.Score", "G2M.Score"),
                     verbose = F)
```

### PCA

PCA works by transforming the original gene expression data into a new set of orthogonal variables, called principal components (PCs). These PCs are linear combinations of the original genes and are ordered such that the first PC captures the maximum amount of variance in the data, followed by the second PC capturing the maximum remaining variance orthogonal to the first PC, and so on.

```{r}
scRNA_brain_seurat_QC <- RunPCA(scRNA_brain_seurat_QC, features = VariableFeatures(object = scRNA_brain_seurat_QC))
```


Elbow plot is used to determine the optimal number of components or clusters to use in the analysis.

```{r}
ElbowPlot(scRNA_brain_seurat_QC, ndims = 50)
```


```{r}
scRNA_brain_seurat_QC <- RunUMAP(scRNA_brain_seurat_QC, dims = 1:15, verbose = F)
```


```{r}
##checking metadata for doublets
head(scRNA_brain_seurat_QC@meta.data)
```

### Removing doublets

```{r}
nExp <- round(ncol(scRNA_brain_seurat_QC) * 0.04)  # expect 4% doublets
scRNA_brain_seurat_QC <- doubletFinder(scRNA_brain_seurat_QC, pN = 0.25, pK = 0.09, nExp = nExp, PCs = 1:15)
# name of the DF prediction can change, so extract the correct column name.

DF.name = colnames(scRNA_brain_seurat_QC@meta.data)[grepl("DF.classification", colnames(scRNA_brain_seurat_QC@meta.data))]
```


```{r}
DimPlot(scRNA_brain_seurat_QC) + NoAxes()
DimPlot(scRNA_brain_seurat_QC, group.by = DF.name) + NoAxes()
```


```{r}
VlnPlot(scRNA_brain_seurat_QC, features = "n_genes", group.by = DF.name, pt.size = 0.1)
```



```{r}
scRNA_brain_seurat_QC = scRNA_brain_seurat_QC[, scRNA_brain_seurat_QC@meta.data[, DF.name] == "Singlet"]
dim(scRNA_brain_seurat_QC)
```


```{r}
##save data
saveRDS(scRNA_brain_seurat_QC, file = "../data/scRNA_brain_seurat_QC_fix.rds")
```


```{r}
sessionInfo()
```

