---
title: "Tutorial 5. Transcription factor activity analysis with scRNA-seq"
author: "Iguaracy Souza and Umran YAMAN"
date: 'Compiled: `r Sys.Date()`'  
output: 
  html_document:
    toc: TRUE
    code_folding:  "show"
    theme: united
    df_print: kable
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  warning = FALSE,
  error = TRUE,
  echo = TRUE,
  message = FALSE,
  fig.width = 8,
  time_it = TRUE
)
```

### Data and library import

Load necessary libraries, and read in the preprocessed scRNA-seq dataset.


```{r}
suppressPackageStartupMessages({
  library("Seurat")
  library("dplyr")
  library("dorothea")
  library("readr")
  library("purrr")
  library("decoupleR")
 })
```

```{r}
scRNA_brain <- readRDS("../data/scRNA_brain_seurat_QC.rds")
```

### Identify differential expressed genes across conditions


Calculate the average expression by cell type:

```{r}
# Calculate the average expression by cell type
avg_exp <- AverageExpression(object = scRNA_brain, assays = "RNA",slot = "counts", group.by = c("major.celltype", "Pathology"))

# Extract the expression matrix
exp_matrix <- avg_exp$RNA

# Apply log1p transformation
log_avg_exp <- log1p(exp_matrix)

# Check the first few entries of the log-transformed data
print(head(log_avg_exp))

# Calculate the sum of expression values for each gene
gene_sums <- rowSums(log_avg_exp)
# Plot a histogram of the row sums
hist(gene_sums, breaks = 100, main = "Distribution of Gene Expression Sums")
```


### Filter out the lowly expressed genes

```{r}
# Set a minimum expression threshold
min_exp_threshold <- 1

# Filter out the low expressed genes
filtered_log_avg_exp <- log_avg_exp[gene_sums > min_exp_threshold, ]
filt_gene_sums <- rowSums(filtered_log_avg_exp)
hist(filt_gene_sums, breaks = 100, main = "Distribution of Gene Expression Sums")
```


### Load dorothea regulons

```{r}
Regulon_file <- read_csv("../data/Regulon_file_get_collectri.csv")
```


### Run wmean

```{r}
scRNA_brain_acts <- run_wmean(mat=filtered_log_avg_exp, net=Regulon_file, .source='source', .target='target',
                                 .mor='mor', times = 1000, minsize = 25)

scRNA_brain_acts <- scRNA_brain_acts %>%
  filter(statistic == 'norm_wmean', p_value < 0.05)
scRNA_brain_acts
```


Now you can check the TF activity per condition: 

```{r}
# Filter rows for early_AD
early_AD <- scRNA_brain_acts %>%
  filter(grepl("earlyAD", condition))

# Filter rows for non_AD
non_AD <- scRNA_brain_acts %>%
  filter(grepl("nonAD", condition))

# Filter rows for lateAD
lateAD <- scRNA_brain_acts %>%
  filter(grepl("lateAD", condition))

write_csv(early_AD, "../results/scRNA_brain_earlyAD_collectTRI.csv")
write_csv(non_AD, "../results/scRNA_brain_nonAD_collectTRI.csv")
write_csv(lateAD, "../results/scRNA_brain_lateAD_collectTRI.csv")
```

